compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -486,7 +486,7 @@ trait TypeAssigner {
     tree.withType(defn.NothingType)
 
   def assignType(tree: untpd.WhileDo)(implicit ctx: Context): WhileDo =
-    tree.withType(defn.UnitType)
+    tree.withType(if (tree.cond eq EmptyTree) defn.NothingType else defn.UnitType)
```

Comment:
Why does this occur ?

Reply:
Because we want a way to represent an infinite `while` loop, whose type is therefore `Nothing`. It's the whole point of the second commit in this PR.